packages/keystone/adapterCache.js

```diff
@@ -0,0 +1,335 @@
+
+            if (cached) {
+                const cacheLastUpdate = cached.lastUpdate
+                if (cacheLastUpdate && cacheLastUpdate === tableLastUpdate) {
+                    middleware.cacheHits++
+                    const cacheEvent = middleware.getCacheEvent({
+                        type: 'HIT',
+                        table: listName,
+                        key,
+                        result: JSON.stringify(cached.response),
+                    })
+                    middleware.writeChangeToHistory({ cache, event: cacheEvent, table: listName } )
+                    middleware.logEvent({ event: cacheEvent })
+                    return cloneDeep(cached.response)
+                }
+            }
+
+            response = await originalItemsQuery.apply(listAdapter, [args, opts] )
+
+            let copiedResponse = cloneDeep(response)
+
+            cache[listName][key] = {
+                lastUpdate: tableLastUpdate,
+                response: copiedResponse,
+            }
+
+            const cacheEvent = middleware.getCacheEvent({
+                type: 'MISS',
+                key,
+                table: listName,
+                result: JSON.stringify(copiedResponse),
+            })
+            middleware.writeChangeToHistory({ cache, event: cacheEvent, table: listName } )
+            middleware.logEvent({ event: cacheEvent })
+
+            return response
+        }
+
+        const originalUpdate = listAdapter.update
+        listAdapter.update = patchAdapterFunction(listName, 'UPDATE', originalUpdate, listAdapter, middleware)
+
+        const originalCreate = listAdapter.create
+        listAdapter.create = patchAdapterFunction(listName, 'CREATE', originalCreate, listAdapter, middleware)
+
+        const originalDelete = listAdapter.delete
+        listAdapter.delete = patchAdapterFunction(listName, 'DELETE', originalDelete, listAdapter, middleware)
+    }
+}
+
+/**
+ * Patches a keystone mutation to add cache functionality
+ * @param {string} listName
+ * @param {string} functionName
+ * @param {function} f
+ * @param {{}} listAdapter
+ * @param {AdapterCacheMiddleware} cache
+ * @returns {function(...[*]): Promise<*>}
+ */
+function patchAdapterFunction ( listName, functionName, f, listAdapter, cache ) {
+    return async ( ...args ) => {
+        const functionResult = await f.apply(listAdapter, args )
+
+        if (cache.debugMode) {
+            const cacheEvent = cache.getCacheEvent({ type: listName, table: listName })
+            cache.writeChangeToHistory({ cache: cache.cache, event: cacheEvent, table: listName })
+        }
+
+        if (cache.logging) { logger.info(`CREATE: ${functionResult}`) }
+        return functionResult
+    }
+}
+
+/**
+ * Stringifies complex objects with circular dependencies
+ * @param obj
+ * @returns {string}
+ */
+function stringifyComplexObj (obj){
+    const result = {}
+    for (const prop in obj ){
+        if (!obj.hasOwnProperty(prop)) { continue }
```

Comment:
Maybe combine this with `||` operator in a single line?

Reply:
Done